packages/ruleset/src/utils/validate-subschemas.js

```diff
@@ -0,0 +1,72 @@
+// Subschemas include property schemas (for an object schema), item schemas
+// (for an array schema), and applicator schemas (such as those in an allOf,
+// anyOf, or oneOf property), plus all subschemas of those schemas.
+
+const validateSubschemas = (schema, path, validate) => {
+  const errors = [];
+  // invoke validation function
+  errors.push(...validate(schema, path));
+
+  // recursively process subschemas
+  if (schema.properties) {
+    for (const property of Object.entries(schema.properties)) {
+      errors.push(
+        ...validateSubschemas(
+          property[1],
+          [...path, 'properties', property[0]],
+          validate
+        )
+      );
+    }
+  }
+
+  if (schema.items) {
+    errors.push(
+      ...validateSubschemas(schema.items, [...path, 'items'], validate)
+    );
+  }
+
+  if (
+    schema.additionalProperties &&
+    typeof schema.additionalProperties === 'object'
+  ) {
+    errors.push(
+      ...validateSubschemas(
+        schema.additionalProperties,
+        [...path, 'additionalProperties'],
+        validate
+      )
+    );
+  }
+
+  // partial to reduce code duplication between applicator schemas
+  function processComposedSchemas(schemas, whichType) {
+    if (Array.isArray(schemas)) {
+      schemas.forEach((s, i) => {
+        errors.push(
+          ...validateSubschemas(s, [...path, whichType, i], validate)
+        );
+      });
+    }
+  }
+
+  if (schema.allOf) {
+    processComposedSchemas(schema.allOf, 'allOf');
+  }
+
+  if (schema.oneOf) {
+    processComposedSchemas(schema.oneOf, 'oneOf');
+  }
+
+  if (schema.anyOf) {
+    processComposedSchemas(schema.anyOf, 'anyOf');
+  }
+
+  if (schema.not) {
+    processComposedSchemas(schema.not, 'not');
+  }
```

Comment:
```suggestion
  for (const applicatorType of ['allOf', 'oneOf', 'anyOf', 'not']) {
    if (Array.isArray(schema[applicatorType])) {
      schema[applicatorType].forEach((s, i) => {
        errors.push(
          ...validateSubschemas(s, [...path, whichType, i], validate)
        );
      });
    }
  }
```

Reply:
Ah, nice! No good reason. In my head, I wanted each schema checked to be its own "if" statement but it also bothered me that I was doing the same thing four times in a row. I like this much better! There's a small typo in your suggestion so I'm just going to apply the change manually and re-push